Move to `importmap`, `stimulus`, `turbo` and adapt all current JS

nimmolo opened the pull request:
This PR is a reboot of #1583, and reuses most of the work started in #1584.

It requires thorough manual testing to merge.
Currently passing tests. 

> [!IMPORTANT] 
> Be sure to run `bundle` before testing. This PR adds several gems.
>
> Also, please run `./bin/rails assets:clobber`. @mo-nathan and @pellaea report they ran `rake assets:precompile` to get the branch to boot, however `assets: clobber` is preferable: `precompile` will compile the assets once to `public/assets/`, but then you will be [stuck with stale JS files](https://stackoverflow.com/a/74945834/3357635). Apparently in development, once you create this folder, Rails will always give priority to anything in the folder over newer files, and not update them. `rails assets:clobber` deletes the entire folder `public/assets/` (intended for use in production, not development) and restores the automated creation of `digest`-versioned files, which ensure you're loading current code.

> [!NOTE] 
> System tests were seeming flaky, but it turns out i needed to modify the tests. Now fixed. 

---

The features listed below have been converted to `Stimulus`/`Turbo`, and should be working correctly as of now. There should be no change in functionality unless noted.

Where the Javascript has been rewritten:

### Indexes ###
- `lazyload_controller.js` **Lazyloading** of images (not only on indexes, but most noticeable there) should work without issues. This is the JS library that draws gray boxes the right shape and size, and fills them in with photos after the page loads. 
- `matrix-table_controller.js` The **matrix table spacers** are invisible... but if this script is not working, there will be rhythmic gaps in the grid of observations. This is a script needed in Bootstrap 3 only, and it may soon disappear, but the JS had to be converted.
- `lightgallery_controller.js` **Theater view** - brand new lightbox, please check it out and let me know what you think.
> [!NOTE]
> This PR changes the JS library for "theater view" of images. `Lightbox` is not actively maintained. `LightGallery` works similarly, but I'd say it's spiffier, and it has more features that we may use in the future.

### Map Views ###
Completely recoded. 
> [!NOTE]
> Before, we had a `gmap` class that built (in Ruby) the JS required to call the Google Maps API. This JS got built and drew the map on page load. This workflow doesn't work, however, if the API, the page or the map div is loaded asynchronously. 
>
> To make it work, we attach a Stimulus JS "controller" that "observes" the map div or its container, and when that div appears, the controller imports the Google Maps API as a module, and it draws the map entirely in JS, which is considerably simpler than building JS in Ruby. The gmap class, and all its callers, are here removed.

- `map_controller.js` Check **map views** for index queries of observations, locations or names
- **Show/Edit location** map UI: In edit, moving the map should update the inputs, and updating the inputs should move the map. 
NEW: 
  - The **rectangle** delimiting the location bounds is now a native Google Maps `Rectangle` object, not a `Polyline` with draggable points. The UI is the same, though. You can drag the corners or the box.
  - There's a new **"Get elevation" button** to retrieve locations without elevation from the Google Maps `ElevationService`. If you move the rectangle or marker at all, it will automatically look up the high/low elevations for the location and fill the inputs in.

### Show Observation ###
- `thumbnail-map_controller.js` The **thumbnail map interface** has been recoded in `Stimulus`. You can click the thumbnail map and be redirected to a Google map of the obs location, also recoded in `Stimulus` (above).
- `modal-toggle_controller.js` Links to **modal forms** for namings, collection numbers, herbarium records and sequences on the show obs page have been converted to `Turbo`, but they're intercepted by this controller so the response is rendered in a Bootstrap modal, rather than elsewhere in the page. If you're working on a form inside a modal, and close it without submitting, you can re-open the same form again and find it as you left it. On the links themselves, there are some new icons in place of words for CRUD actions. 
- `section-update_controller.js` A successful submit should seamlessly update the relevant section of the page without a reload, and update the page flash. An unsuccessful submit should update a flash message in the modal. To experience all the bells and whistles, please test one of your own obs.
- `specimen_controller.js` simply shows/hides the specimen fields, `collection_number` and `herbarium_record`
- `naming-vote_controller.js` Changing **naming votes** on any show obs page should update the naming table, as currently. (`Stimulus` just submits the `form` on select change, `section-update_controller.js` handles the update.)
- `suggestions_controller.js` **Naming suggestions** `Stimulus` performs the `fetch` call to the python script on the image server, but... this does not return results, as presently. No change ;) 
- **External links** - Changed this UI to reuse the methods and templates of the modal forms. This meant adding a minimal Rails controller with `Turbo` responses, rather than the current `AjaxController` approach. It's the same simple UI, but rendered in a Bootstrap modal rather than a native JS "alert" box, and handled by `Turbo`.
- **Ask observer a question** - moved to an "email" icon beside the observer's name. Triggers a modal box to compose the email.
- `autocompleter_controller.js` **Autocompleters** in all forms are converted to `Stimulus`. 
- `year-input_controller.js` **Year inputs** should all be getting changed to text fields, but this is now a separate `Stimulus` controller from the "autocompleter".

### Create Observation form ###
- `map_controller.js` **Map interface**: If you open the map with either the place name or lat/lng inputs filled, the map should try to find the place and set a marker. If you move the map marker by clicking on the map, it should update the observation's lat/lng/alt inputs as you move the marker.
- `obs-form-images_controller.js` **Image uploads**: big new `Stimulus` controller. Check that uploading images with EXIF data offers dialog boxes to sync date/gps with obs date/location if desired.
> [!NOTE]
> This PR swaps out the JS library that we use to read uploaded image EXIF data from `exif.js`, not actively maintained, to `exif_reader`.

### Translations ###
- `translation_controller.js` **Translations UI** recoded as `Turbo` endpoints for the `edit` and `index` actions, with a `Stimulus` controller to handle form interactions. Should work as presently. Minor note: I simplified the mechanism for showing previous versions of long translations in a JS `alert` box, in favor of printing them inline and unabbreviated below the form. You can scroll to them.

### Donations ###
- `donate_controller.js` **Donations form** should work as presently. Filling in an "other amount" checks the "other amount" radio button. That's all the JS was doing.

### Advanced Search ###
- `advanced-search_controller.js` The UI should work as presently, which is to say "not entirely". The JS concerned in this PR just disables filters for irrelevant models... for example if you are "advanced searching" Names, the filters for Locations are greyed out. This is here moved to a `Stimulus` controller and should work as before.

### Name Lister ###
- `name-list_controller.js` should do exactly the same thing, hopefully. 
> [!CAUTION] 
> The `name_list_data.js` primer of genera, species and names generated by the `refresh_name_lister_cache` script has been moved to the new `app/javascripts` folder and slightly rewritten. 
>
> NOTE: `name_list_data.js` WILL NEED TO BE MANUALLY GENERATED ON PRODUCTION AT DEPLOY TIME. It is regenerated by a chron script daily, though.

### Image exports ###
- I can't find where the `image_export.js` script was ever actually implemented. It is theoretically bound to a button generated by a helper, for **marking an image as exportable** or no. I can't find any view that prints that button, though. I have faithfully reproduced the (defunct)ionality, though. Moved out of `AJAXController` to a `Turbo` `Images::ExportsController`. 
> [!NOTE]
> I'm in a pretty good position to fix exports, if anybody wants to explain to me where the UI is and how it's used, and how it should work. I think @mo-nathan uses it for ML.

### Visual Group Status buttons ###
- Moved this out of the `AJAXController` too, to a new `Turbo` `VisualGroups::ImagesController`, and used a CRUD route `update` to render the `Turbo` response. 

### Photo uploads ###
- `file-input_controller.js` File size validation moved to `Stimulus` for "Select file..." single-image fields on Account profile, Glossary term, Herbarium, Project, and Species list forms.

### Various scraps of JS ###
- `name-panels_controller.js` Show name: the **Classification** and **Lifeform** panels are sized to match each other with a `Stimulus` one-liner
- `naming-reason_controller.js` Create obs/Propose naming: **reasons** section. A JS one-liner focuses the textarea for a reason, when you open it by checking a "reason" box. 
___

### Explanation of the PR ###

The goal here is to bring the good stuff from Rails 7 into 6.1 as gems, and theoretically to make updating all the JS easier.

`Turbo` is the Rails standard way to update the page `AJAX`-style, more quickly than a full page response re-executing all the app code. It has a small Javascript library that intercepts internal links and makes specially-formatted requests to Rails controllers. If a controller gets a `format :turbo_stream` request, it can render HTML for part of the page, and Turbo's JS will replace/update/append that part of the page. We are incorporating it on an `opt-in` basis, meaning we have to set `data-turbo="true"` on form elements and `data-turbo-stream="true"` on link elements, to engage Turbo.

> Just a heads up - this PR implements current Turbo (version 7 on Rails 6). In `Turbo 8`, just released in beta, the step of writing special partial responses for each action, which was necessary with AJAX and simplified with Turbo, becomes completely unnecessary. Turbo 8 allows us to forego/delete these `turbo_stream` responses and just `redirect` to actions rendering the entire page, Rails-4-style. Turbo JS intercepts the request to avoid app re-execution, then examines the response HTML, diffs it with the current DOM, and only updates what has changed. Pretty neat — if this were available when `turbolinks` was first introduced it would have been less confusing to implement. But the `dom-morphing` JS library it uses is quite new. I believe the idea is borrowed from frameworks like React.

`Stimulus` is a standardized way to bind UX JS to page elements which may appear either on page load, or any time thereafter, as with a `Turbo` refresh. (It uses a newer standard JS function called `MutationObserver` which watches changes in the DOM, rather than `document.ready` or `window.onload`.)

To use `Turbo` and `Stimulus`, this PR must change the way Javascript is loaded into the app. The PR retires our `javascript_helpers` including `javascript_include` and `inject_javascript_at_end`, which print JS in the page when the app renders a full page load. With `importmap`, Javascript is initially loaded by Sprockets via a reconfigured `link_tree`, but the important difference is what the `importmap` itself does. 

`importmap` is a new-ish browser standard that provides an index of supplemental `module`s and their urls. The urls provided may be from anywhere: our server, or a CDN. Note that a `module` is now a specific type of object in JS, containing a chunk of JS whose `export`ed methods can be accessed at any point in runtime, as long as they are `imported` by the JS that needs them. When any of our JS asks to `import` a `module`, the browser will check if it's already been loaded, and conditionally load it if needed. 

The Rails `importmap` is configured in `config/importmap.rb` and generated and added to the Rails template by `javascript_import_tags`, a helper method of the `importmap-rails` gem. Note: to modify this config, it's generally better to use the `/bin/importmap` command `pin`, which can optionally `--download` any JS to `vendor/javascript`so we serve it from our own server. Future JS modules will need to be "pinned" to the `importmap` - no more gems for JS modules. Once they're `pin`ned, modules can be imported to `app/javascript/application.js` if they need to be globally loaded, or imported to whichever module or controller needs to use them. 

Our own shared utilities modules live in `app/javascript/src/`. All modules there are automatically pinned and added to the `importmap`, so they can be used by any `Stimulus` controller.

All the JS "sprinkles" on our pages are now `Stimulus` controllers, which are pretty easy to write. The reason it's necessary is that `Stimulus` is able to observe and bind to items on the page that have been dynamically loaded, for example by a `Turbo` partial page load. If we really take advantage of `Turbo` to speed up MO, this could be nearly any element on the page. 

`Stimulus` controllers are assigned to specific DOM elements with the HTML attribute `data-controller`. `Stimulus` watches the DOM for the appearance of these elements, and instantiates a controller for each element that calls for one. Once `connected`, the controller instance has a discrete scope: the specified element and its children. However, controllers can "communicate" with each other by `dispatch`ing custom JS `event`s. We use these events, for example, in page updates to show obs: when a section like `naming` is updated, it dispatches an event to the `window` which the `modal_controller` takes as a signal to close the modal it's attached to. (This way, the modal only closes if the `Turbo` response succeeds; otherwise `Turbo` renders any flash messages in the modal.)

To make use of the less verbose `Turbo` pattern to return partial page updates, and to move away from jQuery as a dependency (even though it's great), this PR retires the `jQuery.ajax` method, in favor of:
- links that do built-in `Turbo` requests to controller actions. To enable this, form elements only need a `data-turbo="true"` attribute, and links (sending `GET` requests) need a `data-turbo-stream="true"` attribute. If those are present, `turboJS` will pick up the submit or click event, and handle the request and response, updating the page without us writing any JS whatsoever. The receiving Rails controller must supply a `turbo_stream` format response in Ruby, containing the HTML to be sent to the page, and specifying the HTML `id` of the element to be `replace`d or `update`d. The template file for this response must not have the file extension `.html.erb`, or Turbo will report `no template found`. The template file extension can be `.turbo_stream.erb` or just `.erb`.
- `Stimulus` controllers making JS `fetch` calls directly to Rails controller actions (these can be `GET` or `POST`), such as for the Autocompleters, with the CSRF token automatically added by the `requestjs-rails` gem
- `Stimulus` controllers making JS `fetch` calls outwards, to services like the `Google Maps API`

> Why not leave our JS in jQuery? 

With partial page reloads (previously via AJAX, and now Turbo), any of our `jQuery` interactions that relied on bindings to the `document.ready` event become unreliable: the document might have been ready a long time ago, and the page might be fundamentally different now. 

As we start to use `Turbo` more, and our page navigation starts to avoid reloading the app from scratch, we need a way to bind our JS UI to the newly loaded page elements. In Javascript, it's a little verbose to do that, but `Stimulus` handles it with a standard, documented interface that's easier to write.

`Importmap` is a newer browser feature that just shows the browser where (what url) to load conditionally-imported modules from. For example, if something on the page calls for a Google Maps API JS function to be loaded, the importmap says where that can be loaded from. The `importmap-rails` gem allows us to use modern JS modules loaded in this way. (The `webpacker` gem was a previous solution for this, but very messy. The adoption of `importmaps` as a web standard has made it a lot easier to implement.) 

`importmap-rails` (or a similar solution to load modules) is required to use most of the newer, actively-maintained JS libraries, including Bootstrap 5.

nimmolo committed:
Merge branch 'main' into nimmo-importmap-stimulus

nimmolo committed:
build out map controller

nimmolo committed:
unnecessary config

nimmolo committed:
Merge branch 'main' into nimmo-importmap-stimulus

nimmolo committed:
Update map_helper.rb

update attr names from new MapSet

nimmolo committed:
Update _map.erb

nimmolo committed:
Move jstz to importmap

nimmolo committed:
tidy up

nimmolo committed:
Matrix table controller

nimmolo committed:
Allow for rearranging

nimmolo committed:
Organize map_helper

nimmolo committed:
Merge branch 'main' into nimmo-importmap-stimulus

nimmolo committed:
sketch out controller a bit

nimmolo committed:
working for not-editable

nimmolo committed:
editable marker/box interface

nimmolo committed:
reorder methods

nimmolo committed:
Use coords for markers

nimmolo committed:
Fixed with simpler, draggable points and rectangles?

nimmolo committed:
Editable sort of working. Eliminate superfluous gmap partial from location form

nimmolo committed:
Figure out map_args

nimmolo committed:
get edit loc fields ready for controller actions

nimmolo committed:
start to add elevation to loc form

nimmolo committed:
Elevation updater works

nimmolo committed:
Input handlers for edit_location

nimmolo committed:
Update obs-form-images_controller.js

Fix an error

nimmolo committed:
Update libraries

nimmolo committed:
Temp Fix autocompleter controller key consts

nimmolo committed:
Update map_controller.js

nimmolo committed:
make a lazyload controller for body element

nimmolo committed:
Fix loading error in map_controller

nimmolo committed:
spiffed up form and map_controller, added get elevations button

Clean up

nimmolo commented:
This is not done yet, and not passing tests yet, but it's a big PR and it may be useful to test features in stages.

The following UI should be converted to Stimulus/Turbo as of now. There should be no change in functionality unless noted:
- Lazyloading of images pretty much everywhere
- The matrix table spacers (Bootstrap 3 only, may soon disappear, but the JS had to be converted)
- Lightbox - NEW LIGHTGALLERY. Should work as well as the old one and look spiffier
- Show obs thumbnail map interface, clicks through to a Google map of obs location
- Modal forms on the show obs page should work as before and seamlessly update the relevant section of the page. There may be some new icons in place of words. Test one of your own obs for all the bells and whistles.
- Naming votes on any show obs page, should update the naming table as presently
- Naming suggestions: AJAX calls to the python script do not work, as presently (no change) but the call is executed ;)
- Autocompleters 
- Create obs form: image uploads, check that image date/gps syncs with obs date/location if desired
- Create obs form: map interface
- Year inputs should all be changed to text fields, but not through the "autocompleter" mechanism
- Check Maps for a query of obs, locations or names
- Show/Edit location: map UI. In edit, moving the map should update the inputs, and updating the inputs should move the map. NEW: There's a new "Get elevation" button for locations without elevation, and in any case if you move the rectangle or marker, it will look up the high/low elevations for the location.
- Translations - the UI should work as presently, but I dismantled the "alert" mechanism for showing previous versions of long translations. They just show inline below the form, and you can scroll to them.

TODO: 
- Advanced search
- Name lister
- Donate
- External links
- Image exports

nimmolo committed:
Scaffold new controllers

nimmolo committed:
advanced search controller

nimmolo committed:
donate controller

nimmolo committed:
move donate.js to pre-trash

nimmolo committed:
External links turbo controller

nimmolo committed:
move external_link to pre-trash

nimmolo committed:
Merge branch 'main' into nimmo-importmap-stimulus

nimmolo committed:
Remove call for javascript_include advanced_search

nimmolo committed:
remove commented out javascript include account_api_keys

nimmolo committed:
new visual groups images controller

nimmolo committed:
Add turbo update partial and test

nimmolo committed:
data-turbo on image links

remo,ve ajax controllers

nimmolo committed:
Clean up map args

nimmolo committed:
Clean up herbaria show

nimmolo committed:
Add check for initialized object when making info window

nimmolo committed:
Fix obs_form_images targets

nimmolo committed:
Fix rounding in obs system test location extents

nimmolo committed:
Fix delete button args form: { data: {} }

nimmolo committed:
Merge branch 'main' into nimmo-importmap-stimulus

nimmolo committed:
Merge branch 'nimmo-importmap-stimulus' into nimmo-importmap-last-controllers

nimmolo committed:
Update _links.html.erb

nimmolo committed:
Move external links controllers and templates under observations

nimmolo committed:
Improved, but still not getting turbo_stream response from external links controller

nimmolo committed:
external link form, actions, translations all sorted

Change ivar from @new_sites to @other_sites, move method `externalsites user can add links to` to the obs model as an instance method

nimmolo committed:
Update content_helper.rb

Help_block can now take a block

nimmolo committed:
Whoops - move it to a helper. TIdy up panel

nimmolo committed:
fix visgroupimg user handling

nimmolo committed:
Fix external sites test user handling

nimmolo committed:
Fix status links not showing

nimmolo committed:
Merge pull request #1695 from MushroomObserver/nimmo-importmap-last-controllers

Last three Turbo/Stimulus conversions

nimmolo committed:
Try an images/exports controller with turbo

nimmolo committed:
Remove javascript_include visgrpstatus

nimmolo committed:
rework name_lister without jq

nimmolo committed:
fix test for image export

nimmolo committed:
namelstr maybe close?

nimmolo committed:
Thoroughly debug var names, add targets

nimmolo committed:
Use targets for NL_DIVS

nimmolo committed:
Update gemfile for new requestjs-rails release

nimmolo committed:
more debugging

nimmolo committed:
Update name-list_controller.js

nimmolo committed:
Move and reformat name_list_data.js

nimmolo committed:
Mostly fix name-list

nimmolo committed:
Update refresh_name_lister_cache

nimmolo committed:
Merge pull request #1696 from MushroomObserver/nimmo-importmap-last-controllers

Image Export controller with turbo response

nimmolo committed:
Move EVENT_KEYS to utilities as an object

nimmolo committed:
Retire name_lister.js

nimmolo committed:
Tweak some SCSS

nimmolo committed:
Update _name_lister.scss

nimmolo committed:
Update name-list_controller.js

Rename variables for stimulus controller

nimmolo committed:
make form an .erb

nimmolo committed:
make new an erb

nimmolo committed:
Merge pull request #1697 from MushroomObserver/nimmo-importmap-last-controllers

Cleanup name-list controller

nimmolo committed:
change internal consts to UPPER_SNAKE_CASE

nimmolo committed:
More tweaks to name lister

nimmolo committed:
Fix new and special case autocompleters, swap mechanism

nimmolo committed:
Fix bugs in autocompleter and adv serch ctrlr

nimmolo committed:
cleanup

nimmolo committed:
Make file input validation controller

nimmolo committed:
Further clear up manifest

nimmolo committed:
Merge branch 'main' into nimmo-importmap-stimulus

nimmolo committed:
archive js

nimmolo committed:
more archiving

nimmolo committed:
more archive

nimmolo committed:
put jQuery back in manifest

nimmolo committed:
Update year-input_controller.js

data attributes

nimmolo committed:
Update _namings.erb

remove commented lines

nimmolo committed:
Remove pre-trash

nimmolo committed:
Update year-input_controller.js

nimmolo committed:
naming reason input focuser: scaffold but not working

nimmolo committed:
Remove "huge_image" js

mo-nathan commented:
I tried getting this to work but got:

Showing /Users/nathan/src/mushroom-observer/app/views/layouts/application.html.erb where line #13 raised:
undefined local variable or method `javascript_importmap_tags' for #<ActionView::Base:0x000000001234c8>
  '.freeze;@output_buffer.append=( javascript_importmap_tags );@output_buffer.safe_append='

Any idea what might be wrong?

nimmolo committed:
Name panels

nimmolo committed:
add data-stimulus connected

nimmolo committed:
Remove javascript helpers

nimmolo committed:
Remove gmap class for building js in ruby

nimmolo committed:
Update _head.html.erb

Remove ie8 shim

nimmolo committed:
Delete update.js.erb

Unused

nimmolo committed:
Update .rubocop.yml

Remove mention of gmaps class

nimmolo committed:
Update _media_query_tests.html.erb

Update js example

nimmolo commented:
Hi @mo-nathan, thanks for trying it out. You may just need to bundle the gems, and restart the server.

That error means that the `importmap-rails` gem is not loaded. It's a method of that gem.

Note that I've just made a bunch of commits cleaning up the branch, you probably should update locally before you try it again.

nimmolo committed:
Update locations_controller.rb

Make the class legible in VSCode

nimmolo committed:
Merge branch 'main' into nimmo-importmap-stimulus

nimmolo committed:
Remove some TODO, FIXME and comments

nimmolo committed:
Rubocop stuff

nimmolo committed:
Delete update.js.erb

nimmolo committed:
Update name_list_data.js

New fixture

nimmolo committed:
Create observation_show_system_test.rb

nimmolo committed:
regularize obs modal link classes

nimmolo committed:
System test for show obs page

nimmolo committed:
make a namings_helper#new_naming_tab

nimmolo committed:
adj arg names

nimmolo committed:
tab for new_comment_link

nimmolo committed:
Nest external link form params

nimmolo committed:
Fix a whole bunch of naming, image, vote links for modals, exif links for modals

nimmolo committed:
Update namings_helper.rb

nimmolo committed:
System test working, Modal close via turbo

Many things fixed. bullet suggestions

mo-nathan commented:
Overall this looks fantastic!  I think I did spot a couple issues with the lightbox functionality.  It's now possible to get the text to overlap the image which doesn't look great.  If you search for images of Cortinarius and look at one of them (e.g., 533485) and then scroll to one that has a lot more (e.g., 533449) you get an page like the one attached.  The Notes field in particular seems to be the problem.  If the lines of text allowed for notes could get limited to some standard number I think that would be better.  A vertical scrolling panel might be better.  Alternatively putting the Notes in a separate div to the right of the other text might make better use of the space.  I would also keep the left justification of the current implementation (rather than center).
![Uploading Screenshot 2023-11-22 at 7.41.31 AM.png…]()

mo-nathan commented:
Not sure what's going on with the screenshot I tried to upload.  I think you can reproduce the issue just by creating an observation with lots of notes and then starting the lightbox on another observation.

nimmolo committed:
Fill out the obs show system test

nimmolo committed:
comment/callbacks: bullet

nimmolo committed:
fix tests

nimmolo committed:
Truncate observation notes at 150 characters...

nimmolo commented:
@mo-nathan Thanks for trying it out, and for the Lightgallery feedback.  

The format of the lightgallery caption is _very_ configurable, including having it conditionally display to the side or bottom of the image depending on (something), but it will take some time to dive into. What we have here is basically what it renders out of the can, with almost no customization, and I had so much else to do on this PR that I didn't want to get off the trail too far. 

I just added a `truncate` to the observation notes, to guarantee it's not more than 150 characters. But i'm still getting overlap. I'll keep playing with it a bit more.

nimmolo committed:
Fix exif, shorten notes further

nimmolo committed:
Truncate notes differently

nimmolo committed:
Fix "Send observer a question" modal

nimmolo committed:
Spacing on + links

nimmolo committed:
Get send obs email link out of top tabs, does not work there

nimmolo committed:
Update modal-toggle_controller.js

nimmolo committed:
Move obs img uploads to own ctlr

nimmolo committed:
Finish move

nimmolo committed:
Move autocompleters to own controller

nimmolo committed:
cleanup

nimmolo committed:
remove ajax controller

nimmolo committed:
Update autocompleters_controller.rb

Don't need disable_filters

nimmolo committed:
Update autocompleter_controller.js

nimmolo committed:
Merge pull request #1698 from MushroomObserver/nimmo-obs-img-uploads-ctlr

Move observation image upload and autocompleters to their own controllers

nimmolo committed:
Adjust lightgallery caption

nimmolo commented:
Yahoo! Pretty sure I figured out how to limit the caption height to avoid overlap, and get it left aligned (but not to the photo). Easier to read.

nimmolo commented:
@JoeCohen I know you're on a family vacation and probably having fun. 

Would you like me to wait so you can kick the tires on this thing, or do you prefer to skip it?

nimmolo committed:
log the collapse event. debugging

nimmolo committed:
Fix jQuery import, and naming reasons UI

nimmolo committed:
Remove jquery-rails gem

nimmolo committed:
Add coverage for naming vote % table modal, do Bullet optimization

nimmolo committed:
Create help_identify_system_test.rb

JoeCohen commented:
@nimmolo: Please go ahead when you're ready.
If I can finish responding to the comments on Add date range to Projects PR
#1658,
I will look at this PR, but I don't know when I'm able to do that.

On Fri, Nov 24, 2023 at 3:13 PM andrew nimmo ***@***.***>
wrote:

> @JoeCohen <https://github.com/JoeCohen> I know you're on a family
> vacation and probably having fun.
>
> Would you like me to wait so you can kick the tires on this thing, or do
> you prefer to skip it?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/MushroomObserver/mushroom-observer/pull/1683#issuecomment-1826054843>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAALDFCCPD2N3SEN4UZN7JTYGD5WNAVCNFSM6AAAAAA6Z7ASECVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTQMRWGA2TIOBUGM>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

nimmolo committed:
Fix identify UI

Need ID identifiers on modal forms

nimmolo committed:
Fix id and caption helper, system test

nimmolo committed:
Give the modal naming and the modal naming form an obs_id, there may be multiple on page

nimmolo committed:
Fix obs show naming test, remove `logged_in` param from namings table (you must be logged in to even get the table)

nimmolo commented:
I've figured out you guys' secret plan... you're just stalling me out until I actually write system tests and debug this stuff! 

Hah. Well, your plan is working! That's what i've been doing and it is really helpful at finding bugs.

nimmolo committed:
Update observation_show_system_test.rb

nimmolo committed:
rubocop stuff

nimmolo committed:
Switch references of lightbox to lightgallery

nimmolo committed:
Streamlines use of modal and section-update controllers with naming table

mo-nathan commented:
I tried checking this out again and just get gray blocks for images.  The Chrome JavaScript console says:

Uncaught TypeError: Failed to resolve module specifier "jquery3". Relative references must start with either "/", "./", or "../".

nimmolo commented:
@mo-nathan Your branch is out of date. On Nov 26, 2023, at 4:57 AM, Nathan Wilson ***@***.***> wrote:﻿
I tried checking this out again and just get gray blocks for images.  The Chrome JavaScript console says:
Uncaught TypeError: Failed to resolve module specifier "jquery3". Relative references must start with either "/", "./", or "../".

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you were mentioned.Message ID: ***@***.***>

mo-nathan commented:
```% git pull
Already up to date.
% git status
On branch nimmo-importmap-stimulus
Your branch is up to date with 'origin/nimmo-importmap-stimulus'.
```

nimmolo commented:
I hear you, but it can't be, because `jquery3` is not imported anywhere on the branch!

mo-nathan commented:
```
% git grep jquery3
app/javascript/application.js:// import "jquery3"
```

mo-nathan commented:
Yeah, something weird is going on.  I even deleted the branch from my local repo and pulled it again and the file is still there, but in the PR it's not.  Makes no sense.

mo-nathan commented:
I even see it if I delete and then reclone a local repo.

```
(base) nathan@Nathans-MacBook-Pro tmp % rm -rf mushroom-observer 
(base) nathan@Nathans-MacBook-Pro tmp % git clone git@github.com:MushroomObserver/mushroom-observer.git
Cloning into 'mushroom-observer'...
remote: Enumerating objects: 143323, done.
remote: Counting objects: 100% (5511/5511), done.
remote: Compressing objects: 100% (1772/1772), done.
remote: Total 143323 (delta 3916), reused 5193 (delta 3676), pack-reused 137812
Receiving objects: 100% (143323/143323), 84.35 MiB | 28.54 MiB/s, done.
Resolving deltas: 100% (112725/112725), done.
(base) nathan@Nathans-MacBook-Pro tmp % cd mushroom-observer 
(base) nathan@Nathans-MacBook-Pro mushroom-observer % git checkout nimmo-importmap-stimulus
branch 'nimmo-importmap-stimulus' set up to track 'origin/nimmo-importmap-stimulus'.
Switched to a new branch 'nimmo-importmap-stimulus'
(base) nathan@Nathans-MacBook-Pro mushroom-observer % git grep jquery3
app/javascript/application.js:// import "jquery3"
```

mo-nathan commented:
I can even do the clone using this branch and still see it:
```
(base) nathan@Nathans-MacBook-Pro tmp % git clone -b nimmo-importmap-stimulus git@github.com:MushroomObserver/mushroom-observer.git
Cloning into 'mushroom-observer'...
remote: Enumerating objects: 143323, done.
remote: Counting objects: 100% (5511/5511), done.
remote: Compressing objects: 100% (1772/1772), done.
remote: Total 143323 (delta 3915), reused 5195 (delta 3676), pack-reused 137812
Receiving objects: 100% (143323/143323), 84.38 MiB | 29.73 MiB/s, done.
Resolving deltas: 100% (112717/112717), done.
(base) nathan@Nathans-MacBook-Pro tmp % cd mushroom-observer 
(base) nathan@Nathans-MacBook-Pro mushroom-observer % git grep jquery3    
app/javascript/application.js:// import "jquery3"
```

nimmolo commented:
It doesn't work like Sprockets where a magic comment like `//= link mo_utilities` is effectively an "import".

mo-nathan commented:
Deleting the entire file still has the same problem even though `git grep` cannot find jquery3 any more.  Tried restarting Chrome and tried Safari.  Is there something I'm supposed do similar to `bundle install` but for JavaScript stuff?

nimmolo commented:
Let's continue on Slack!

nimmolo committed:
Remove Geocoder class for LocationsController#new, use Stimulus map_controller

nimmolo committed:
Fix info windows for new Google Maps rectangles

nimmolo committed:
Remove references to Geocoder

nimmolo committed:
Remove outdated reference to `caption` local in obs info partial

nimmolo committed:
Fix "email observer question" link logic in helper

nimmolo committed:
Add new string extension to strip newlines from Textile, for lightbox caption

nimmolo committed:
Consolidate obs show info/details partials into details

nimmolo committed:
Whoops, left debug in

nimmolo committed:
Let a geocoded location be a bounded rectangle! Adjust tests accordingly

nimmolo commented:
@mo-nathan - one UI that i believe only you can accurately review is Visual Group Status.

nimmolo committed:
Update observation_form_system_test.rb

nimmolo committed:
Update application_system_test_case.rb

Increase default max wait time for system tests on CI

nimmolo committed:
Update observation_form_system_test.rb

nimmolo committed:
Update links in translations. whew

nimmolo committed:
Change all references to account/prefs

nimmolo committed:
map_controller: DRY and fix drawMarker

nimmolo committed:
Update map_controller.js

Renamings in preparation for maybe consolidating with obs-form-map

nimmolo committed:
Update observation_show_system_test.rb

nimmolo committed:
comments

nimmolo committed:
Update map_controller.js

comments

nimmolo committed:
Update observation_show_system_test.rb

Jeez, that test is ornery.

nimmolo committed:
Update observation_show_system_test.rb

Move sleep

nimmolo committed:
Reinstate progress whirly for naming votes, adj test

nimmolo committed:
Add default "turbo-submits-with" data attribute to submit_button helper

nimmolo committed:
Merge branch 'main' into nimmo-importmap-stimulus

nimmolo committed:
Comments: make modal ids more specific

Also - use div not turbo-frame on namings and comments
Still need better system tests for comment create/update

nimmolo committed:
Fix modal IDs. (Add record ID to the html ID for edit form modals)

Also adjusts tests

nimmolo commented:
@mo-nathan @JoeCohen -  just fixed some issues Jason pointed out with the modal forms on show-obs. 

I will definitely not merge this until I'm sure you have both tested the UI out pretty thoroughly. There are plenty of corners of MO where I do not know the expected behavior.

mo-nathan commented:
Bug in the edit map interface.
1) Go to the edit location page for any existing location
2) Click on the text box for any compass direction and select all the text.
3) Type in a value that includes a decimal point (e.g., "41.7").
4) Causes the values in the box you're entering into to go crazy.  I've seen them jump "90" or "180" or "18" or "-180".  Note, you can edit an existing value as long as you don't type in the character '.'.  I also saw the altitude values get reset to NaN just by typing '.' in one of the other other fields.

mo-nathan commented:
Bug in the create map workflow for a new observation.  I wanted to create an observation in "Scotland, St. Andrews, St. Andrews Botanic Garden".  It found a reasonable lat/long for this location, but did not create a surrounding rectangle.  It also didn't update the map when I entered values for that rectangle, and when I tried to save the value it complained about getting NaN values in the altitude values (which it had filled in).  Going back to the page and getting it to figure out the altitude info allowed me to save the location.  After enough futzing I was able to get it to save and have the rectangle (North 56.338, West -2.81, East -2.8, South 56.333).  However, this is still not completely accurate, but when I went into to edit it, the pins weren't visible on the edit window so I couldn't drag and drop them.

nimmolo commented:
@mo-nathan - Belatedly realizing you're talking about the create **Location** workflow, the one that is initiated after creating an Observation in a place that's not in our Locations table yet. 

This does indeed have a bounding box, and it is **not the bounding box of the observation collection location, it's the bounding box of a registered location in the database**, which I believe the user does not expect to be able to edit.

This is to me a major UI issue with our maintaining a Locations table for planet Earth. It's not clear to the user that they're creating two objects in the database.  (Who but Igor Safonov expects us to be maintaining this kind of info manually? Judging from the way my friends have forgotten how to even drive to Berkeley Bowl without being instructed by phone GPS, I think users are inclined to accept geolocation data at face value.)

nimmolo commented:
Regarding the "decimal point input bug": I have seen this on previous commits, but I'm not getting it currently, in Chrome, Firefox or Safari.

I suspect your branch might be out of date. (I think this was fixed at least a week ago.)

JoeCohen commented:
Edit Location:
As soon as I click Edit Location it overrides the preexisting N, S, E, W.
Example: 
- Go to http://localhost:3000/locations/6420
Observe that the boundaries are: 
North: 45.3806°
West: -122.742° East: -122.696°
South: 45.3374°
- click Edit Location
Result: 
North:
45.4025342
West: -122.7273726 East: -122.6475383
South:
45.3560451

JoeCohen commented:
Edit Location, resizing the box displays the resized Elevations as `NaN`
<img width="266" alt="Screenshot 2023-12-02 at 4 36 15 PM" src="https://github.com/MushroomObserver/mushroom-observer/assets/45460/2015ab83-5041-4f31-9e37-66264e35ecf7">

JoeCohen commented:
Or maybe make them the equivalent of bold?
> Show Observation Could you make the Proposed Names Edit and Destroy icons larger, or use text instead? At first, I thought there was no longer a way to Edit or Destroy a Proposed Name. It took me a couple of minutes to figure it out. I think we'll get questions about this.

JoeCohen commented:
PS: compare the Proposed Names glyphicons for Edit and Destroy to same glyphicons for Image.
At least on my screen the Image ones are bigger, bolder, clearer:
<img width="717" alt="Screenshot 2023-12-02 at 4 59 50 PM" src="https://github.com/MushroomObserver/mushroom-observer/assets/45460/864886ae-5d26-4c2d-8d37-e01457f59723">

nimmolo committed:
Fix map_controller's getElevations

JoeCohen commented:
Cannot change an Observation's default Image
- Edit an Observation which has > 1 image
- Change the default by clicking on a radio button
- Save Edits
Result: default Image is unchanged.

nimmolo commented:
Good question about the API billing. The main difference here from current is that Jason used Ruby to build the JS that calls the Google Maps API — but every call still counts as a call. The only new thing I added is looking up the Elevations.

We do three [billable things with Google Maps](https://developers.google.com/maps/documentation/javascript/usage-and-billing) (they actually call them SKUs, like Safeway does).

[SKU: Dynamic Maps](https://developers.google.com/maps/documentation/javascript/usage-and-billing#dynamic-maps)
User interactions with the map, such as panning, zooming, or switching map layers, do not generate additional map loads. 
No change in usage from current. Every call to Google to draw a map for a set of coordinates counts as a call.

[SKU: Elevation](https://developers.google.com/maps/documentation/elevation/usage-and-billing#elevation)
Our code currently does some elevation calls, but this PR enables checking them more easily and does them by default on locations. I think that counts as an improvement, but it will increase usage of this item.
 
[SKU: Geocoding](https://developers.google.com/maps/documentation/geocoding/usage-and-billing#geocoding)
We do this currently via the Geocode class. No anticipated change in usage.

nimmolo committed:
Fix setting default image in edit observation

nimmolo committed:
Remove debugger

nimmolo committed:
Fix image drop zone (whole form)

nimmolo committed:
Bump up naming CRUD icon size

nimmolo commented:
@JoeCohen - I've fixed the image drag and drop on the form, I believe. Thank you for catching it. (This is not ever going to be easy to system test, since it requires orchestrating mouse movements with a file picked up outside the browser. It may not even be possible.)

nimmolo committed:
Add instruction to click on map for create obs

nimmolo committed:
Remove debugger lines

nimmolo committed:
Fix uploads_controller (was dropping session)

nimmolo committed:
Update observation_form_system_test.rb

Somehow save_with_log(@naming) isn't finishing,
ObservationsController::NewAndCreate line 176

nimmolo committed:
Use request.js for obs form images

nimmolo committed:
Update application_controller.rb

Remove ajax after_action

JoeCohen commented:
Send a Comment/Report a Bug 
- http://localhost:3000/emails/ask_webmaster_question
Result: fills in `Your email address` with webmaster@mushroomobserver.org instead of the user's email.

nimmolo commented:
Bullet is suddenly complaining about an unoptimized query, and I need help. I believe it's an erroneous report, and the action that's causing it (`ObservationsController#create`) is already told to `skip_bullet` around itself, so i cannot imagine how Bullet could be "more" turned off than it is. 

To reproduce, pull the current branch and run:
```
rails t test/system/observation_form_system_test.rb:78
```

Bullet wants us to eager load the `:rss_log` on an observation that has just been created (and logged). There is no Observation query anywhere in the trace of methods being called, only the model instance method `HerbariumRecord.observations` that pulls the observations associated with an herbarium record, and this is an AR automatically-created method. So... it kind of has to be a false positive? Other than that, we're just passing the new observation around to the methods that create associated records for c_n, h_r, etc.

The message is:
```
Minitest::UnexpectedError:         Bullet::Notification::UnoptimizedQueryError: user: nimmo
        POST http://localhost:3000/observations?approved_name=&approved_where=USA%2C+California%2C+Pasadena
        USE eager loading detected
          Observation => [:rss_log]
          Add to your query: .includes([:rss_log])
        Call stack
          /Users/nimmo/Documents/GitHub/developer-startup/mushroom-observer/app/models/abstract_model.rb:661:in `log'
          /Users/nimmo/Documents/GitHub/developer-startup/mushroom-observer/app/models/herbarium_record.rb:101:in `add_observation'
          /Users/nimmo/Documents/GitHub/developer-startup/mushroom-observer/app/controllers/observations_controller/new_and_create.rb:226:in `save_herbarium_record'
          /Users/nimmo/Documents/GitHub/developer-startup/mushroom-observer/app/controllers/observations_controller/new_and_create.rb:169:in `save_everything_else'
          /Users/nimmo/Documents/GitHub/developer-startup/mushroom-observer/app/controllers/observations_controller/new_and_create.rb:104:in `create'
          /Users/nimmo/Documents/GitHub/developer-startup/mushroom-observer/app/controllers/application_controller.rb:137:in `skip_bullet'
          /Users/nimmo/Documents/GitHub/developer-startup/mushroom-observer/app/controllers/application_controller.rb:233:in `catch_errors_and_log_request_stats'
```

mo-nathan committed:
Allow bullet to be turned off multiple times and only go back on when it has been turned back on the right number of times

mo-nathan committed:
Store the Bullet.n_plus_one_query_enable values on the stack

nimmolo committed:
Merge pull request #1703 from MushroomObserver/njw-bullet-depth

Save bullet state on the stack in skip_bullet

nimmolo committed:
Update obs-form-images_controller.js

Fix and annotate the POST fetchRequest

nimmolo committed:
Update observation_form_system_test.rb

Fix typo in regex!

nimmolo committed:
Update observation_form_system_test.rb

nimmolo committed:
Update _collection_numbers.erb

Fix icon layout

nimmolo committed:
Update uploads_controller.rb

layout false

nimmolo committed:
Update obs-form-images_controller.js

Fix submit interruption

nimmolo committed:
Update observation_show_system_test.rb

Increase wait for naming vote to save

nimmolo committed:
Update observation_show_system_test.rb

nimmolo committed:
Fix form data-turbo attributes

Not form turbo: true but data: { turbo: true }

nimmolo committed:
Better comments, wait for naming modal in autocompleter test

nimmolo commented:
Two remaining questions in my mind:
- Create Location — @mo-nathan's question boils down to this: Should we coerce a Location to be a rectangle, in the frequent cases where Google Maps API returns a point (a postal address), as it does for St. Andrew's Botanic Garden? The current UI does force rectangles, by intentionally fuzzing the point into a box. My (not too strong) opinion is, that's sort of polluting the data, if we're considering Google's API the source of truth, but it's not too hard to implement. However, every time the place-name is geolocated with the API, it will revert to being a point. (My other opinion is that if we drop the Locations table and its many dependencies, using the Google Maps API to get geocodes, and send those to the db to return observations/names for a place, this app may be easier to maintain and use. I'm not sure this would work with the API, but the API may not even permit that kind of query. Mobile could make its own calls to the API.) 
- Does VisualGroup Status UI work as expected?

mo-nathan commented:
Fuzzing the point and keeping the Location object seems like the simplest thing to do.  While I would support switching to real locations that seems like a substantial amount of work that it would be better to do in a separate PR.  There a bunch of interesting questions there like what do we do about places where Google disagrees with data that humans have entered.  In the case of the botanic garden, for example, the point it gives is not the point for where I made the collection I want to enter.  I expect we will NOT want to lose any human entered and reviewed data, but rather introduce the Google data as another type of location.

mo-nathan commented:
The VisualGroup UI does not work as expected.  In particular, the "Needs Review" selection does not work correctly.  It should delete any associated VisualGroupImage.  Instead it seems to cause it to set VisualGroupImage.included to false.  I also noticed that the image voting has started throwing errors like:
```
Template is missing
Missing partial images/votes/_update with {:locale=>[:en], :formats=>[:turbo_stream], :variants=>[], :handlers=>[:raw, :erb, :html, :builder, :ruby, :jbuilder]}. Searched in: * "/Users/nathan/src/mushroom-observer/app/views" * "/Users/nathan/.gem/ruby/3.1.4/gems/turbo-rails-1.5.0/app/views"
```

nimmolo commented:
@mo-nathan You don't expect missing partials on a @nimmolo PR?

nimmolo committed:
Start to converge stimulus maps

nimmolo committed:
Switch points to rectangle

nimmolo committed:
continue consolidating controllers

nimmolo committed:
Change MapSet to store lng, not long, adjust controller

nimmolo committed:
Make attr lng accessible on observatoions passed to CollapsibleCollection, fix tests

nimmolo committed:
Continue controller work

nimmolo committed:
Fix image voting

mo-nathan commented on app/controllers/visual_groups/images_controller.rb:
```diff
@@ -0,0 +1,40 @@
+      vgi = visual_group.visual_group_images.find_by(image_id: image_id)
+      included = (@status == "true")
+      if @status == ""
+        vgi&.destroy
```
The problem is when you click 'Needs Review'.  It should just get selected (become dark and bold), but on your branch it selects 'Exclude' instead.  I haven't checked what it does on the backend.  It is intentional that the page doesn't otherwise change so you can power through lots of images without stuff jumping around.  It also makes it more forgiving if you make a mistake.

nimmolo replied:
Got it, tx for the quick response.

nimmolo replied:
Ah, the issue is that private method `update_visual_group_image` is coercing the value of `status` to a boolean, when it needs to preserve the `nil` value for `needs_review`. 

The method doesn't seem any different from `main`... will debug further.

nimmolo replied:
Found my bug. Should an image marked "Needs Review" still appear in the visual group on reload, and only the "Excluded" images fall out of the group? 

On `main` the "Needs Review" images do not appear in the group on reload, but then i'm wondering how you would eventually "review" those images if they don't appear in the group.

(I missed the coercion to string of `params[:value].to_s` so i was checking for `""` instead of `nil`.)

nimmolo replied:
Thank you, that helps!

However, on `main`, just now, I do *not* find the "Needs Review" images on the Edit view.
I *do* find the "Excluded" images when I click that link, and also the "Included". But even when returning to "Needs Review" via the link that appears on "Excluded", no images appear. 

I know you've been using this UI, and it must work, but that doesn't seem right. I notice that in `AjaxController::VisualGroupStatus`, `vgi&.destroy` is only called on a VisualGroupImage with status == "", i.e. "Needs Review". Can you confirm this is correct?

nimmolo replied:
Ok as far as I can tell there are two things that keep VisualGroupImages from behaving as described. One is that the status values map like this:
```
needs_review | included | excluded
nil                         true           false
```
but having a `needs_review` value is the one that gets the VGI deleted, where I would expect it to be the `excluded` value.
```ruby
    if @value == ""
      vgi&.destroy
      status = nil
```

The second issue is that the VisualGroupImages table does not allow nil values on the `included` column.
```
    t.boolean "included", default: true, null: false
```

nimmolo replied:
@mo-nathan - Just want to be sure you saw this question about the setup of `VisualGroupImages`. No rush, but when you have time to consider, please let me know what you think. 

I think the controller behavior should maybe be changed.

nimmolo replied:
Resolved via email - this is a bug on `main` and will leave as-is on this PR.

nimmolo committed:
VisualGroups::ImagesController @status->status

nimmolo committed:
Update images_controller.rb

remove debugger

nimmolo committed:
Fix nil handling in VisualGroups::ImagesController

nimmolo committed:
generalize map controller

nimmolo committed:
image status - params are always strings

nimmolo committed:
Fix findOnMap and zoom on click

nimmolo committed:
Map obs form too eager to fill lat/lng

nimmolo committed:
Update map_controller.js

Try to debug. It's not hitting makeMapClickable()

nimmolo committed:
Reconfigure map to not load on obs form until clicked

Capybara acting up

nimmolo committed:
Adjust translations

nimmolo committed:
Update assets.rb

Remove obsolete instruction (?)

nimmolo committed:
geo-coordinates-parser

nimmolo committed:
Obs form system test working again

nimmolo committed:
Fix obs_show_system_test (controller action create dropping User)

nimmolo committed:
Try various things to fix cookie bug

nimmolo committed:
Whoa, did this solve it?

nimmolo committed:
Fix reload of naming table (must reload @obs)

nimmolo committed:
Update observation_show_system_test.rb

nimmolo committed:
Merge branch 'main' into nimmo-importmap-stimulus

nimmolo committed:
Bullet - Add counter cache to Name - observations_count

nimmolo committed:
Merge branch 'main' into nimmo-importmap-stimulus

nimmolo committed:
Update shared_private_methods.rb

nimmolo committed:
Redo refresh_primer logic a bit

nimmolo committed:
Tweak var names

nimmolo committed:
Remove high-maintenance counter cache from obs.name

nimmolo committed:
Update species_lists_controller.rb

Skip bullet on create also

nimmolo committed:
rubocop stuff

nimmolo committed:
Fix genus/species matches

nimmolo committed:
Update observation_show_system_test.rb

Reinstate auto-complete delay on naming form

nimmolo committed:
Update autocompleter_controller.js

Use requestjs for autocompleter

nimmolo committed:
Fix bullet exception

Grant exception for Bullet on this association and remove ineffective `:skip_bullet` call

nimmolo committed:
Update observation_show_system_test.rb

Use auto-complete to slow down the test

nimmolo committed:
Update species_lists_controller.rb

nimmolo committed:
Update observation_show_system_test.rb

Scrolling to maybe fixes click_link?

nimmolo committed:
Merge branch 'main' into nimmo-importmap-stimulus

nimmolo committed:
Merge branch 'main' into nimmo-importmap-stimulus
